### PR TITLE
Use coroutines instead of xpcall in typechecker

### PR DESCRIPTION
**Pull request by hugomg**
I think that the coroutine version is easier to understand, and is more
idiomatic Lua. The stack traces look better and it doesn't need that "special
metatable" trick.

(Obrigado pela sugestão, Ana)

### CODE REVIEW

**Comment by hugomg** [comment]
To compare the stack traces I added an `error("boom!")` call inside the `Exp.Integer` case of `check_exp`.

Here is what the new coroutine version of the stack trace looks like

```
./pallene/checker.lua:156: ./pallene/checker.lua:506: boom!
stack traceback:
	[C]: in function 'error'
	./pallene/checker.lua:506: in upvalue 'check_exp'
	./pallene/checker.lua:225: in upvalue 'check_top_level'
	./pallene/checker.lua:162: in upvalue 'check_program'
	./pallene/checker.lua:142: in function <./pallene/checker.lua:141>

stack traceback:
	./pallene/checker.lua:156: in function <./pallene/checker.lua:140>
	(...tail calls...)
	./pallene/driver.lua:42: in function <./pallene/driver.lua:27>
	(...tail calls...)
	spec/checker_spec.lua:8: in upvalue 'run_checker'
	spec/checker_spec.lua:30: in function <spec/checker_spec.lua:29>
```

This is how the old xpcall-based stack traces looked like:

```
/pallene/checker.lua:155: ./pallene/checker.lua:506: boom!
stack traceback:
	[C]: in function 'error'
	./pallene/checker.lua:506: in upvalue 'check_exp'
	./pallene/checker.lua:225: in upvalue 'check_top_level'
	./pallene/checker.lua:162: in function <./pallene/checker.lua:160>
	[C]: in function 'xpcall'
	./pallene/checker.lua:148: in function <./pallene/checker.lua:147>
	(...tail calls...)
	./pallene/driver.lua:42: in function <./pallene/driver.lua:27>
	(...tail calls...)
	spec/checker_spec.lua:8: in upvalue 'run_checker'
	spec/checker_spec.lua:30: in function <spec/checker_spec.lua:29>
	[C]: in function 'xpcall'
	...
	/home/hugo/.luarocks/share/lua/5.3/busted/core.lua:312: in field 'execute'
	/home/hugo/.luarocks/share/lua/5.3/busted/block.lua:155: in field 'execute'
	/home/hugo/.luarocks/share/lua/5.3/busted/init.lua:7: in upvalue 'executor'
	/home/hugo/.luarocks/share/lua/5.3/busted/core.lua:312: in function </home/hugo/.luarocks/share/lua/5.3/busted/core.lua:312>
	[C]: in function 'xpcall'
	/home/hugo/.luarocks/share/lua/5.3/busted/core.lua:178: in field 'safe'
	/home/hugo/.luarocks/share/lua/5.3/busted/core.lua:312: in field 'execute'
	/home/hugo/.luarocks/share/lua/5.3/busted/execute.lua:58: in local 'execute'
	/home/hugo/.luarocks/share/lua/5.3/busted/runner.lua:197: in function 'busted.runner'
	/home/hugo/.luarocks/lib64/luarocks/rocks-5.3/busted/2.0.rc13-0/bin/busted:3: in main chunk
	[C]: in ?

stack traceback:
	./pallene/checker.lua:155: in function <./pallene/checker.lua:147>
	(...tail calls...)
	./pallene/driver.lua:42: in function <./pallene/driver.lua:27>
	(...tail calls...)
	spec/checker_spec.lua:8: in upvalue 'run_checker'
	spec/checker_spec.lua:30: in function <spec/checker_spec.lua:29>
```

File: pallene/checker.lua [hunk 1]
```diff
@@ -145,15 +138,22 @@ end
 --
 
 type_check = function(prog_ast)
-    local ok, err = xpcall(check_program, debug.traceback, prog_ast)
+    local co = coroutine.create(function()
```
Comment (gligneul) [inline pallene/checker.lua]: Why not?
```suggestion
    local co = coroutine.create(check_program)
    local ok, err_msg = coroutine.resume(co, prog_ast)
```

File: pallene/checker.lua [hunk 2]
```diff
@@ -145,15 +138,22 @@ end
...
+    end)
+    local ok, err_msg = coroutine.resume(co)
     if ok then
-        return prog_ast, {} -- TODO:  no {}
-    else
-        if getmetatable(err) == TypeError then
-            return false, { err.msg } -- TODO  no {}
+        if coroutine.status(co) == "dead" then
+            -- User's program passed type checker
+            return prog_ast, {} -- TODO:  no {}
```
Comment (gligneul) [inline pallene/checker.lua]: I'd rather remove this TODO. This is non-informative.

File: pallene/checker.lua [hunk 3]
```diff
@@ -145,15 +138,22 @@ end
...
         else
-            error(err)
+            -- User's program has a type error
+            return false, { err_msg }
```
Comment (gligneul) [inline pallene/checker.lua]:
```suggestion
            return false, {err_msg}
```

File: pallene/checker.lua [hunk 4]
```diff
@@ -145,15 +138,22 @@ end
...
+        check_program(prog_ast)
+    end)
+    local ok, err_msg = coroutine.resume(co)
     if ok then
-        return prog_ast, {} -- TODO:  no {}
-    else
-        if getmetatable(err) == TypeError then
-            return false, { err.msg } -- TODO  no {}
+        if coroutine.status(co) == "dead" then
+            -- User's program passed type checker
```
Comment (gligneul) [inline pallene/checker.lua]: The code seems very straight forward, I would remove this comments.

File: pallene/checker.lua (hunk 2 above)
Comment (hugomg) [reply pallene/checker.lua]: Agreed, the issue tracker is a better place for this kind of thing.

File: pallene/checker.lua (hunk 4 above)
Comment (hugomg) [reply pallene/checker.lua]: I don't think it is *that* straightforward. For example, "err_msg" means different things in each of the branches, and the distinction between type error in the compiler and type error in the user code is inherently confusing. Let's keep the comments :)

**Pull request merged**